content/courses/advanced-ada/parts/resource_management/anonymous_access_types.rst

```diff
@@ -1958,12 +1960,559 @@ access-to-subprogram types can be solved by declaring them as anonymous types.
+that we've seen before apply to anonymous access-to-subprograms as well. Also,
+we had a discussion about
+:ref:`accessibility rules and access-to-subprograms <Adv_Ada_Accessibility_Rules_Access_To_Subprograms>`
+in the previous chapter. In this section, we review some of the rules that we
+already know and discuss how they relate to anonymous access-to-subprograms.
 
-    Complete section!
+.. admonition:: In the Ada Reference Manual
+
+    - :arm22:`3.10 Access Types <3-10>`
+
+
+Named vs. anonymous access-to-subprograms
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Let's see an example of a named access-to-subprogram type:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Resource_Management.Anonymous_Access_Types.Accessibility_Rules_Anonymous_Access_To_Subprograms.Simple_Example_Named
+    :class: ada-expect-compile-error
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Access_To_Subprogram_Error is
+
+       type PI is access
+         procedure (I : in out Integer);
+
+       P : PI;
+
+       I : Integer := 0;
+    begin
+       declare
+          procedure Add_One (I : in out Integer) is
+          begin
+             I := I + 1;
+          end Add_One;
+       begin
+          P := Add_One'Access;
+       end;
+    end Show_Access_To_Subprogram_Error;
+
+In this example, we get a compilation error because the lifetime of the
+:ada:`Add_One` procedure is shorter than the access type :ada:`PI`.
+
+In contrast, using an anonymous access-to-subprogram type eliminates the
+compilation error, i.e. the assignment :ada:`P := Add_One'Access` becomes
+legal:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Resource_Management.Anonymous_Access_Types.Accessibility_Rules_Anonymous_Access_To_Subprograms.Simple_Example_Anonymous
+    :class: ada-run-expect-failure
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Access_To_Subprogram_Error is
+       P : access procedure (I : in out Integer);
+
+       I : Integer := 0;
+    begin
+       declare
+          procedure Add_One (I : in out Integer) is
+          begin
+             I := I + 1;
+          end Add_One;
+       begin
+          P := Add_One'Access;
+          --  ERROR: Add_One is out-of-scope after
+          --         this line.
+       end;
+    end Show_Access_To_Subprogram_Error;
+
+In this case, the compiler introduces an accessibility check, which fails at
+runtime because the lifetime of :ada:`Add_One` is shorter than the lifetime of
+the access object :ada:`P`.
+
+
+Named vs. anonymous access-to-subprograms as parameters
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Using anonymous access-to-subprograms as parameters allows us to pass
+subprograms at any level. For certain applications, the restrictions that are
+applied to named access types might be too strict, so using anonymous
+access-to-subprograms might be a good way to circumvent those restrictions.
```

Comment:
They also allow the component developer to be independent of the clients' specific access types.

Reply:
Sounds good, I'll add this phrase verbatim.